# 파트세미나 5주차_ 데이터 전처리 및 가공 1 #

## 01. mpg.csg를 불러와서 mpg.csv에서 displ이 4 이하인 자동차와 5 이상인 자동차 중 어떤 자동차의 hwy 평균이 더 높은지 알아보기 ##

### 1-1. mpg.csv 불러오기 ###

In [3]:
import pandas as pd
mpg = pd.read_csv('mpg.csv')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
0           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
1           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
2           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
3           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
4           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
229   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
230   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
231   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
232   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
233   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

    category  
0    compact  
1    compact  
2    compact  
3    compact  
4    compact  
..       ...  
229  midsize  
230  midsize  
231  midsize  
232  midsize  
233  midsize  

[234 rows x 11 columns]

*→ 234개의 행과 11개의 열(변수)로 이루어진 데이터 생성*

### 1-2. 조건에 해당하는 데이터 추출하여 데이터 만들기

In [6]:
displ4 = mpg.query('displ <= 4')
displ5 = mpg.query('displ >= 5')

In [9]:
displ4

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
0           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
1           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
2           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
3           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
4           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
229   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
230   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
231   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
232   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
233   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

    category  
0    compact  
1    compact  
2    compact  
3    compact  
4    compact  
..       ...  
229  midsize  
230  midsize  
231  midsize  
232  midsize  
233  midsize  

[163 rows x 11 columns]

In [10]:
displ5

manufacturer                   model  displ  year  cyl       trans drv  \
18     chevrolet      c1500 suburban 2wd    5.3  2008    8    auto(l4)   r   
19     chevrolet      c1500 suburban 2wd    5.3  2008    8    auto(l4)   r   
20     chevrolet      c1500 suburban 2wd    5.3  2008    8    auto(l4)   r   
21     chevrolet      c1500 suburban 2wd    5.7  1999    8    auto(l4)   r   
22     chevrolet      c1500 suburban 2wd    6.0  2008    8    auto(l4)   r   
23     chevrolet                corvette    5.7  1999    8  manual(m6)   r   
24     chevrolet                corvette    5.7  1999    8    auto(l4)   r   
25     chevrolet                corvette    6.2  2008    8  manual(m6)   r   
26     chevrolet                corvette    6.2  2008    8    auto(s6)   r   
27     chevrolet                corvette    7.0  2008    8  manual(m6)   r   
28     chevrolet         k1500 tahoe 4wd    5.3  2008    8    auto(l4)   4   
29     chevrolet         k1500 tahoe 4wd    5.3  2008    8    auto(l4)   4   
30     chevrolet         k1500 tahoe 4wd    5.7  1999    8    auto(l4)   4   
31     chevrolet         k1500 tahoe 4wd    6.5  1999    8    auto(l4)   4   
55         dodge       dakota pickup 4wd    5.2  1999    8  manual(m5)   4   
56         dodge       dakota pickup 4wd    5.2  1999    8    auto(l4)   4   
61         dodge             durango 4wd    5.2  1999    8    auto(l4)   4   
62         dodge             durango 4wd    5.7  2008    8    auto(l5)   4   
63         dodge             durango 4wd    5.9  1999    8    auto(l4)   4   
70         dodge     ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4   
71         dodge     ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4   
72         dodge     ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4   
73         dodge     ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4   
75          ford          expedition 2wd    5.4  1999    8    auto(l4)   r   
76          ford          expedition 2wd    5.4  2008    8    auto(l6)   r   
82          ford            explorer 4wd    5.0  1999    8    auto(l4)   4   
88          ford         f150 pickup 4wd    5.4  1999    8    auto(l4)   4   
89          ford         f150 pickup 4wd    5.4  2008    8    auto(l4)   4   
98          ford                 mustang    5.4  2008    8  manual(m6)   r   
128         jeep      grand cherokee 4wd    5.7  2008    8    auto(l5)   4   
129         jeep      grand cherokee 4wd    6.1  2008    8    auto(l5)   4   
134      lincoln           navigator 2wd    5.4  1999    8    auto(l4)   r   
135      lincoln           navigator 2wd    5.4  1999    8    auto(l4)   r   
136      lincoln           navigator 2wd    5.4  2008    8    auto(l6)   r   
140      mercury         mountaineer 4wd    5.0  1999    8    auto(l4)   4   
153       nissan          pathfinder 4wd    5.6  2008    8    auto(s5)   4   
158      pontiac              grand prix    5.3  2008    8    auto(s4)   f   
199       toyota  land cruiser wagon 4wd    5.7  2008    8    auto(s6)   4   

     cty  hwy fl    category  
18    14   20  r         suv  
19    11   15  e         suv  
20    14   20  r         suv  
21    13   17  r         suv  
22    12   17  r         suv  
23    16   26  p     2seater  
24    15   23  p     2seater  
25    16   26  p     2seater  
26    15   25  p     2seater  
27    15   24  p     2seater  
28    14   19  r         suv  
29    11   14  e         suv  
30    11   15  r         suv  
31    14   17  d         suv  
55    11   17  r      pickup  
56    11   15  r      pickup  
61    11   16  r         suv  
62    13   18  r         suv  
63    11   15  r         suv  
70    11   15  r      pickup  
71    11   16  r      pickup  
72    13   17  r      pickup  
73    11   15  r      pickup  
75    11   17  r         suv  
76    12   18  r         suv  
82    13   17  r         suv  
88    11   15  r      pickup  
89    13   17  r      pickup  
98    14   20  p  subcompact  
128   13   18  r         suv  


*→ 배기량이 4이하인 자동차, 5이상인 자동차의 데이터를 따로 생성*

### 1-3. 각 데이터의 hwy 평균 확인 ### 

In [11]:
displ4['hwy'].mean()

np.float64(25.96319018404908)

In [ ]:
displ5['hwy'].mean()

np.float64(18.07894736842105)

*→ displ4 hwy평균은 약 25.96이고  displ5의 hwy평균은 약 18.08이므로 배기량 4이하 자동차의 고속도로 연비 평균이 더 높다.*

## 02. audi와 toyota 중 어느 manufactuer의 city 평균이 높은지 알아보기 ##

### 2-1. 조건 해당하는 행 추출하여 데이터 만들기 ### 

In [14]:
audi = mpg.query('manufacturer == "audi"')
toyota = mpg.query('manufacturer == "toyota"')

In [15]:
audi

manufacturer       model  displ  year  cyl       trans drv  cty  hwy fl  \
0          audi          a4    1.8  1999    4    auto(l5)   f   18   29  p   
1          audi          a4    1.8  1999    4  manual(m5)   f   21   29  p   
2          audi          a4    2.0  2008    4  manual(m6)   f   20   31  p   
3          audi          a4    2.0  2008    4    auto(av)   f   21   30  p   
4          audi          a4    2.8  1999    6    auto(l5)   f   16   26  p   
5          audi          a4    2.8  1999    6  manual(m5)   f   18   26  p   
6          audi          a4    3.1  2008    6    auto(av)   f   18   27  p   
7          audi  a4 quattro    1.8  1999    4  manual(m5)   4   18   26  p   
8          audi  a4 quattro    1.8  1999    4    auto(l5)   4   16   25  p   
9          audi  a4 quattro    2.0  2008    4  manual(m6)   4   20   28  p   
10         audi  a4 quattro    2.0  2008    4    auto(s6)   4   19   27  p   
11         audi  a4 quattro    2.8  1999    6    auto(l5)   4   15   25  p   
12         audi  a4 quattro    2.8  1999    6  manual(m5)   4   17   25  p   
13         audi  a4 quattro    3.1  2008    6    auto(s6)   4   17   25  p   
14         audi  a4 quattro    3.1  2008    6  manual(m6)   4   15   25  p   
15         audi  a6 quattro    2.8  1999    6    auto(l5)   4   15   24  p   
16         audi  a6 quattro    3.1  2008    6    auto(s6)   4   17   25  p   
17         audi  a6 quattro    4.2  2008    8    auto(s6)   4   16   23  p   

   category  
0   compact  
1   compact  
2   compact  
3   compact  
4   compact  
5   compact  
6   compact  
7   compact  
8   compact  
9   compact  
10  compact  
11  compact  
12  compact  
13  compact  
14  compact  
15  midsize  
16  midsize  
17  midsize

In [16]:
toyota

manufacturer                   model  displ  year  cyl       trans drv  \
173       toyota             4runner 4wd    2.7  1999    4  manual(m5)   4   
174       toyota             4runner 4wd    2.7  1999    4    auto(l4)   4   
175       toyota             4runner 4wd    3.4  1999    6    auto(l4)   4   
176       toyota             4runner 4wd    3.4  1999    6  manual(m5)   4   
177       toyota             4runner 4wd    4.0  2008    6    auto(l5)   4   
178       toyota             4runner 4wd    4.7  2008    8    auto(l5)   4   
179       toyota                   camry    2.2  1999    4  manual(m5)   f   
180       toyota                   camry    2.2  1999    4    auto(l4)   f   
181       toyota                   camry    2.4  2008    4  manual(m5)   f   
182       toyota                   camry    2.4  2008    4    auto(l5)   f   
183       toyota                   camry    3.0  1999    6    auto(l4)   f   
184       toyota                   camry    3.0  1999    6  manual(m5)   f   
185       toyota                   camry    3.5  2008    6    auto(s6)   f   
186       toyota            camry solara    2.2  1999    4    auto(l4)   f   
187       toyota            camry solara    2.2  1999    4  manual(m5)   f   
188       toyota            camry solara    2.4  2008    4  manual(m5)   f   
189       toyota            camry solara    2.4  2008    4    auto(s5)   f   
190       toyota            camry solara    3.0  1999    6    auto(l4)   f   
191       toyota            camry solara    3.0  1999    6  manual(m5)   f   
192       toyota            camry solara    3.3  2008    6    auto(s5)   f   
193       toyota                 corolla    1.8  1999    4    auto(l3)   f   
194       toyota                 corolla    1.8  1999    4    auto(l4)   f   
195       toyota                 corolla    1.8  1999    4  manual(m5)   f   
196       toyota                 corolla    1.8  2008    4  manual(m5)   f   
197       toyota                 corolla    1.8  2008    4    auto(l4)   f   
198       toyota  land cruiser wagon 4wd    4.7  1999    8    auto(l4)   4   
199       toyota  land cruiser wagon 4wd    5.7  2008    8    auto(s6)   4   
200       toyota       toyota tacoma 4wd    2.7  1999    4  manual(m5)   4   
201       toyota       toyota tacoma 4wd    2.7  1999    4    auto(l4)   4   
202       toyota       toyota tacoma 4wd    2.7  2008    4  manual(m5)   4   
203       toyota       toyota tacoma 4wd    3.4  1999    6  manual(m5)   4   
204       toyota       toyota tacoma 4wd    3.4  1999    6    auto(l4)   4   
205       toyota       toyota tacoma 4wd    4.0  2008    6  manual(m6)   4   
206       toyota       toyota tacoma 4wd    4.0  2008    6    auto(l5)   4   

     cty  hwy fl category  
173   15   20  r      suv  
174   16   20  r      suv  
175   15   19  r      suv  
176   15   17  r      suv  
177   16   20  r      suv  
178   14   17  r      suv  
179   21   29  r  midsize  
180   21   27  r  midsize  
181   21   31  r  midsize  
182   21   31  r  midsize  
183   18   26  r  midsize  
184   18   26  r  midsize  
185   19   28  r  midsize  
186   21   27  r  compact  
187   21   29  r  compact  
188   21   31  r  compact  
189   22   31  r  compact  
190   18   26  r  compact  
191   18   26  r  compact  
192   18   27  r  compact  
193   24   30  r  compact  
194   24   33  r  compact  
195   26   35  r  compact  
196   28   37  r  compact  
197   26   35  r  compact  
198   11   15  r      suv  
199   13   18  r      suv  
200   15   20  r   pickup  
201   16   20  r   pickup  
202   17   22  r   pickup  
203   15   17  r   pickup  
204   15   19  r   pickup  
205   15   18  r   pickup  
206   16   20  r   pickup

*→ 자동차 제조회사가 audi인 자동차, toyota인 자동차의 데이터를 따로 생성*

### 2-2. 각 데이터의 cty 평균 확인 ###

In [18]:
audi['cty'].mean()

np.float64(17.61111111111111)

In [19]:
toyota['cty'].mean()

np.float64(18.529411764705884)

*→ audi의 cty평균은 약 17.6이고 toyota의 cty평균은 약 18.5이므로 자동차 제조회사가 toyota인 자동차들의 도시 연비 평균이 더 높다.*

## 03. category, city 변수를 추출해 새로운 데이터 mpg_new 생성하고 확인 ##

### 3-1. mpg_new 생성하기 ### 

In [21]:
mpg_new = mpg[['category', 'cty']]
mpg_new

category  cty
0    compact   18
1    compact   21
2    compact   20
3    compact   21
4    compact   16
..       ...  ...
229  midsize   19
230  midsize   21
231  midsize   16
232  midsize   18
233  midsize   17

[234 rows x 2 columns]

*→ category, cty 두 개의 변수만 있는 데이터 생성 (234개의 행)*

### 3-2. head()로 출력하여 확인하기 ### 

In [22]:
mpg_new.head()

category  cty
0  compact   18
1  compact   21
2  compact   20
3  compact   21
4  compact   16

*→ head()를 이용해 일부 5개만 출력하여 보다 간편하게 확인 가능*

---------

### 1-2. 

*→  *